packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,398 @@
+        const metadata = await this.#getMetadata(assetsWithoutMetadata);
+
+        const newMetadata = {
+          ...this.state.metadata,
+          ...metadata.assets,
+        };
+        this.update((state) => {
+          state.metadata = newMetadata;
+        });
+      }
+      this.update((state) => {
+        state.allNonEvmTokens[account.id] = assets;
+      });
+    }
+  }
+
+  /**
+   * Handles changes when a new account has been removed.
+   *
+   * @param accountId - The new account id being removed.
+   */
+  async #handleOnAccountRemoved(accountId: string): Promise<void> {
+    const selectedAccounts = this.messagingSystem.call(
+      'AccountsController:listMultichainAccounts',
+    );
+
+    const nonEvmAccounts = selectedAccounts.filter((account) =>
+      this.#isNonEvmAccount(account),
+    );
+    const account: InternalAccount | undefined = nonEvmAccounts.find(
+      (multichainAccount) => multichainAccount.id === accountId,
+    );
+    if (!account) {
+      return;
+    }
+
+    this.update((state) => {
+      delete state.allNonEvmTokens[accountId];
+    });
+  }
+
+  #getAllSnaps(): Snap[] {
+    return this.messagingSystem.call('SnapController:getAll') as Snap[];
+  }
+
+  #getSnapsPermissions(
+    origin: string,
+  ): SubjectPermissions<PermissionConstraint> {
+    return this.messagingSystem.call(
+      'PermissionController:getPermissions',
+      origin,
+    ) as SubjectPermissions<PermissionConstraint>;
+  }
+
+  async #getAssets(
+    accountId: string,
+    snapId: string,
+  ): Promise<CaipAssetType[]> {
+    return await this.#getAssetsList(snapId, accountId);
+  }
+
+  /**
+   * Gets a `KeyringClient` for a Snap.
+   *
+   * @param snapId - ID of the Snap to get the client for.
+   * @param accountId - ID of the account to get the assets for.
+   * @returns A `KeyringClient` for the Snap.
+   */
+  // TODO: update this to use the snap handler
+  async #getAssetsList(
+    snapId: string,
+    accountId: string,
+  ): Promise<CaipAssetType[]> {
+    const result = (await this.messagingSystem.call(
+      'SnapController:handleRequest',
+      {
+        snapId: snapId as SnapId,
+        origin: 'metamask',
+        handler: HandlerType.OnRpcRequest,
+        request: {
+          id: '4dbf133d-9ce3-4d3f-96ac-bfc88d351046',
+          jsonrpc: '2.0',
+          method: 'listAccountAssets',
+          params: {
+            id: accountId,
+          },
+        },
+      },
+    )) as CaipAssetType[];
+
+    return result;
+  }
+
+  // TODO: update this function to get metadata from the snap
+  async #getMetadata(assets: CaipAssetType[]): Promise<AssetLookupResponse> {
+    console.log('🚀 ~ #getMetadata ~ assets:', assets);
+    return Promise.resolve({
+      assets: {
+        'solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1/slip44:501': {
+          name: 'Solana',
+          symbol: 'SOL',
+          native: true,
+          fungible: true,
+          iconBase64:
```

Comment:
shouldn't this be an url?  We are getting the url from the static api

Reply:
I thought so too, it was defined as `iconBase64` in this [document](https://github.com/MetaMask/SIPs/pull/154/files) which i think is different than a url.

Reply:
IMO, this is better to use a base64 here, in case the logo we want to use is not hosted "anywhere".

Also, using "untrusted URLs" seems always """risky""" (I don't have good example for now) 🙈

Reply:
This was updated to be url and now using types from snaps-sdk 🙏